ray-curator/ray_curator/stages/function_decorators.py

```diff
@@ -0,0 +1,120 @@
+"""Utility decorators for creating ProcessingStage instances from simple functions.
+
+This module provides a :func:`processing_stage` decorator that turns a plain
+Python function into a concrete :class:`ray_curator.stages.base.ProcessingStage`.
+
+Example
+-------
+
+```python
+from ray_curator.stages.resources import Resources
+from ray_curator.stages.function_decorators import processing_stage
+
+
+@processing_stage(name="WordCountStage", resources=Resources(cpus=1.0), batch_size=1)
+def word_count(task: SampleTask) -> SampleTask:
+    # Add a *word_count* column to the task's dataframe
+    task.data["word_count"] = task.data["sentence"].str.split().str.len()
+    return task
+```
+
+The variable ``word_count`` now holds an *instance* of a concrete
+``ProcessingStage`` subclass that can be added directly to a
+:class:`ray_curator.pipeline.Pipeline`.
+"""
+
+from __future__ import annotations
+
+import inspect
+from typing import TYPE_CHECKING, TypeVar, cast, overload
+
+from ray_curator.stages.base import ProcessingStage
+from ray_curator.stages.resources import Resources
+from ray_curator.tasks import Task
+
+if TYPE_CHECKING:
+    from collections.abc import Callable
+
+# Type variables representing the Task in/out types handled by the
+# user-provided function.  They must both be (sub-classes of) Task so that the
+# generated ProcessingStage satisfies the base class contract.
+TIn = TypeVar("TIn", bound=Task)
+TOut = TypeVar("TOut", bound=Task)
+
+
+@overload
+def processing_stage(
+    *,
+    name: str,
+    resources: Resources | None = None,
+    batch_size: int | None = None,
+) -> Callable[[Callable[[TIn], TOut | list[TOut]]], ProcessingStage[TIn, TOut]]: ...
+
+
+def processing_stage(
+    *,
+    name: str,
+    resources: Resources | None = None,
```

Comment:
let's also allow `dict[str, float]` so that users don't have to import resources, and they can just do 
```
@processing_stage(name="WordCountStage", resources={"cpus" : 1}, batch_size=1)
def word_count(...)
```

Reply:
I like the suggestion for resources, will add.

IMO `ray_stage_spec`, `xenna_stage_spec`, `inputs`, and `outputs` is a bit overkill. If the user wants control over those things, they can implement a `ProcessingStage` directly instead of using the decorator. Does this sound reasonable?